sdks/java/core/src/main/java/org/apache/beam/sdk/testing/BigqueryMatcher.java

```diff
@@ -144,23 +148,31 @@ Bigquery newBigqueryClient(String applicationName) {
         .build();
   }
 
+  @Nonnull
   @VisibleForTesting
   QueryResponse queryWithRetries(Bigquery bigqueryClient, QueryRequest queryContent,
                                  Sleeper sleeper, BackOff backOff)
       throws IOException, InterruptedException {
-    IOException lastException = null;
+    IOException lastException;
     do {
       try {
-        return bigqueryClient.jobs().query(projectId, queryContent).execute();
+        QueryResponse response = bigqueryClient.jobs().query(projectId, queryContent).execute();
+        if (response != null) {
+          return response;
+        } else {
+          lastException =
+              new IOException("Expected valid response from query job, but received null.");
+        }
       } catch (IOException e) {
         // ignore and retry
         LOG.warn("Ignore the error and retry the query.");
```

Comment:
This will be logged even when you're not ignoring the exception - if this is your last attempt, you'll log "Ignoring", and then immediately throw an exception caused by this one.

Also, this should be more precise (while you're here) - something like:
`LOG.warn("Retrying query (blah) after exception", e)`